Implement persistence for `db_object_import_rule` resource

Tener opened the pull request:
Implement persistence for `db_object_import_rule` resource, defined in https://github.com/gravitational/teleport/pull/37795, using wrappers from https://github.com/gravitational/teleport/pull/37803.

This is a PR in a series of PRs implementing [RFD 151](https://github.com/gravitational/teleport/blob/master/rfd/0151-database-permission-management.md).

Tener committed:
Implement persistence for db_object_import_rule resource

Tener commented:
The PR depends on:
- https://github.com/gravitational/teleport/pull/37795
- https://github.com/gravitational/teleport/pull/37803

The single commit outside of those is https://github.com/gravitational/teleport/pull/37806/commits/612dac827f7fc4117690e7348fd139494f4719ac. I'll rebase the PR against master once the PRs above are merged.

To see how the change looks like in its entirety, please see: https://github.com/gravitational/teleport/pull/35034.

The capstone PR will carry the changelog entry.

Tener commented:
There is just a single dependent PR which isn't merged to master yet: https://github.com/gravitational/teleport/pull/37803.

I've rebased this PR against the latest version of that branch, and changed the merge branch to it as well. This means the PR is now truthful to its real scope and hopefully will be automatically retargeted to master once that PR gets merged.

bl-nero commented on lib/services/databaseobjectimportrule.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+// DatabaseObjectImportRule is the DatabaseObjectImportRule service
+type DatabaseObjectImportRule interface {
```
The name of this interface is exactly the same as the name of type of object that it operates on. This is confusing. I suggest using a plural name: `DatabaseObjectImportRules`, similar to the `Databases` interface in the same package.

Tener replied:
Done

bl-nero commented on lib/services/databaseobjectimportrule.go:
```diff
@@ -0,0 +1,97 @@
+	dbobjectimportrulev1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1"
+)
+
+// DatabaseObjectImportRule is the DatabaseObjectImportRule service
```
```suggestion
// DatabaseObjectImportRules manages DatabaseObjectImportRule resources.
```

bl-nero commented on lib/services/databaseobjectimportrule.go:
```diff
@@ -0,0 +1,97 @@
+	UpdateDatabaseObjectImportRule(ctx context.Context, rule *dbobjectimportrulev1.DatabaseObjectImportRule) (*dbobjectimportrulev1.DatabaseObjectImportRule, error)
+
+	// ListDatabaseObjectImportRules will list DatabaseObjectImportRule resources.
+	ListDatabaseObjectImportRules(ctx context.Context, size int, token string) ([]*dbobjectimportrulev1.DatabaseObjectImportRule, string, error)
```
What does the token do?

Tener replied:
`token` is a standard argument for `ListFoo` RPCs; it tells you where to resume the listing.

Tener replied:
I've renamed it `pageToken` to convey additional context, I hope this works well enough.

bl-nero commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,194 @@
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// Backend encapsulates required backend methods.
```
```suggestion
// Backend interface for manipulating DatabaseObjectImportRule resources.
```

bl-nero commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,194 @@
+	logger     logrus.FieldLogger
+}
+
+// GetDatabaseObjectImportRule gets a DatabaseObjectImportRule by name. It will throw an error if the DatabaseObjectImportRule does not exist.
```
```suggestion
// GetDatabaseObjectImportRule gets a DatabaseObjectImportRule by name. It will return an error if the DatabaseObjectImportRule does not exist.
```
Also, this second sentence should also be included in the interface documentation, especially that we are returning a pointer.

bl-nero commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,194 @@
+	}, nil
+}
+
+// CreateDatabaseObjectImportRule creates a new DatabaseObjectImportRule. It will throw an error if the DatabaseObjectImportRule already
```
```suggestion
// CreateDatabaseObjectImportRule creates a new DatabaseObjectImportRule. It will return an error if the DatabaseObjectImportRule already
```
Here, I think that the interface also would benefit from having this contract explicitly stated in the documentation.

bl-nero commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,194 @@
+	backend Backend,
+	rule *pb.DatabaseObjectImportRule,
+) error {
+	if err := databaseobjectimportrule.ValidateDatabaseObjectImportRule(rule); err != nil {
```
Oops, missing authorization code?

Tener replied:
Indeed, thanks for catching this. I've added the checks now, although I still need to add test coverage.

Tener replied:
@bl-nero I've added the test; should be good now, kindly check.

bl-nero commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := authz.AuthorizeAdminAction(ctx, authCtx); err != nil {
```
Why don't we perform this call in all of the other methods?

Tener replied:
We do now - for those that are in scope.

bl-nero commented on lib/services/databaseobjectimportrule_test.go:
```diff
@@ -0,0 +1,72 @@
+		})},
+	}
+
+	for _, tt := range tests {
```
It's only one test, I don't think we need this loop.

bl-nero commented on tool/tctl/common/resource_command.go:
```diff
@@ -586,6 +588,38 @@ func (rc *ResourceCommand) createBot(ctx context.Context, client auth.ClientI, r
+
+	if existingRule != nil {
+		if !rc.force {
+			return trace.AlreadyExists("rule %q already exists", ruleName)
```
If `CreateDatabaseObjectImportRule` already returns an error if rule already exists, why don't we use it? Otherwise, we have a race condition here and there is a case where we don't create an object even if `force` is specified (If a Get operation returns nothing, then something appears in the backend, and only then we call the Create RPC.)

bl-nero reviewed (changes requested):
I'm requesting changes here, because I suspect some missing authorization calls.

Tener committed:
Apply suggestions from code review

Co-authored-by: Bartosz Leper <bartosz.leper@goteleport.com>

Tener committed:
Use plural interface name: DatabaseObjectImportRules. Fix typos. Refactor test.

Tener committed:
token => pageToken

Tener committed:
Simplify resource command

Tener commented on api/client/client.go:
```diff
@@ -3067,6 +3068,26 @@ func (c *Client) DeleteAllDatabaseServices(ctx context.Context) error {
+	var out []*dbobjectimportrulev1.DatabaseObjectImportRule
+	req := &dbobjectimportrulev1.ListDatabaseObjectImportRulesRequest{}
+	for {
+		resp, err := c.DatabaseObjectImportRuleClient().ListDatabaseObjectImportRules(ctx, req)
```
Good call.

Tener committed:
Fix permission checks

greedy52 commented on api/client/client.go:
```diff
@@ -3067,6 +3068,27 @@ func (c *Client) DeleteAllDatabaseServices(ctx context.Context) error {
 }
 
+// GetDatabaseObjectsImportRules retrieves all database object import rules.
+func (c *Client) GetDatabaseObjectsImportRules(ctx context.Context) ([]*dbobjectimportrulev1.DatabaseObjectImportRule, error) {
```
nit:
```suggestion
func (c *Client) GetDatabaseObjectImportRules(ctx context.Context) ([]*dbobjectimportrulev1.DatabaseObjectImportRule, error) {
```
Should `DatabaseObjects` be plural?

greedy52 replied:
Don't we use `DatabaseObjectImportRule` or like `ListDatabaseObjectImportRules` elsewhere?

bl-nero replied:
I agree with @greedy52 on this one. I'm not a native speaker, but I find it more idiomatic to name a list of object an "object list". Like grocery list, guest list. A pack of wolves is a wolf pack, not a wolves pack. Here's some useful discussion: https://www.proz.com/forum/linguistics/245800-noun_list_is_noun_singular_or_plural.html

Tener replied:
Done.

greedy52 commented on lib/services/databaseobjectimportrule.go:
```diff
@@ -0,0 +1,97 @@
+	dbobjectimportrulev1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1"
+)
+
+// DatabaseObjectImportRules manages DatabaseObjectImportRule resources.
+type DatabaseObjectImportRules interface {
```
Are we caching the rules in database service?

greedy52 replied:
sounds good 👍

greedy52 commented on tool/tctl/common/collection.go:
```diff
@@ -1164,6 +1165,29 @@ func (c *botCollection) writeText(w io.Writer, verbose bool) error {
+	return resources
+}
+
+func (c *databaseObjectImportRuleCollection) writeText(w io.Writer, verbose bool) error {
+	t := asciitable.MakeTable([]string{"Name"})
```
Could we add some more details on the rule in addition to  "Name"?

Tener replied:
Done.

Tener committed:
add details

Tener committed:
refactor, rebase fixes

Tener committed:
better name

Tener committed:
Fix several issues.

- Change type of `db_labels` proto field to the new shared `label` message.
- Implement missing method.
- Cause compilation errors in case of unipmlemented methods.
- Add test coverage for `tctl` commands.
- Change output of Marshal/Unmarshal methods.

Tener committed:
Fix test

greedy52 commented on api/types/label/convert_test.go:
```diff
@@ -0,0 +1,102 @@
+				{Name: "key1", Values: []string{"value1", "value2"}},
+				{Name: "key1", Values: []string{"value3"}},
+			},
+			want: map[string][]string{"key1": {"value1", "value2", "value3"}},
```
nice

greedy52 commented on api/types/label/convert_test.go:
```diff
@@ -0,0 +1,102 @@
+			want: []*labelv1.Label{{Name: "key1", Values: []string{"value1", "value2"}}},
+		},
+		{
+			name: "map with multiple entries",
+			in: map[string][]string{
+				"key1": {"value1", "value2"},
+				"key2": {"value3"},
+			},
+			want: []*labelv1.Label{
+				{Name: "key1", Values: []string{"value1", "value2"}},
+				{Name: "key2", Values: []string{"value3"}},
+			},
```
If you run this -count=100, it will most likely fail. Do we care to sort to make sure this function always give the same order? Or maybe have another version like `FromMapSorted` that sorts it.

Tener replied:
I don't think we care, I've switched to order-invariant comparison in test.

greedy52 commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,222 @@
+		return nil, trace.BadParameter("backend service is required")
+	}
+	if cfg.Logger == nil {
+		cfg.Logger = logrus.WithField(trace.Component, "DatabaseObjectImportRule.service")
```
IIRC, the component gets cut off after several characters?

greedy52 replied:
Some examples from my journalctl log: `[DB:SERVIC]`, `[UPLOAD]`. `[PROXY:AGE]`. I don't think it's a big deal.

Tener replied:
I've trimmed the name somewhat.

greedy52 commented on lib/auth/dbobjectimportrule/dbobjectimportrulev1/service.go:
```diff
@@ -0,0 +1,222 @@
+	return rule, nil
+}
+
+func (rs *DatabaseObjectImportRuleService) UpsertDatabaseObjectImportRule(ctx context.Context, req *pb.UpsertDatabaseObjectImportRuleRequest) (*pb.DatabaseObjectImportRule, error) {
```
nit: go docs

Tener replied:
One day we'll have a linter for this 😅 thanks, fixed.

greedy52 reviewed (approved):
LGTM and tested `tctl`

Tener committed:
Fix nondeterminism in test

Tener committed:
Add mising docs

Tener committed:
Resolve protobuf breaking changes

Tener committed:
Shorten component name for logging

Tener committed:
Add access test

Tener committed:
Fix test

Tener committed:
Update error message and tests

Tener committed:
Merge branch 'master' into tener/rfd-151-part-5-services

Tener committed:
Remove temp test `TestCreateUser`

r0mant reviewed (approved):
Bot.
